test/testdrive/expected_group_size_tuning.td

```diff
@@ -0,0 +1,280 @@
+        ORDER BY l2.l_extendedprice DESC
+        LIMIT 1)
+  FROM lineitem l1
+  GROUP BY l1.l_partkey;
+
+> CREATE MATERIALIZED VIEW lineitem_by_suppkey AS
+  SELECT l1.l_suppkey,
+        MAX(l1.l_extendedprice),
+        (SELECT l2.l_quantity FROM lineitem l2
+        WHERE l2.l_suppkey = l1.l_suppkey
+        ORDER BY l2.l_extendedprice DESC
+        LIMIT 1)
+  FROM lineitem l1
+  GROUP BY l1.l_suppkey;
+
+# Check that we know the values of the perfect hints for TPC-H.
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 AS perfect_hint
+  FROM (
+        SELECT l_orderkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_orderkey
+  );
+15
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_partkey, l_suppkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_partkey, l_suppkey
+  );
+255
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_partkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_partkey
+  );
+255
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_suppkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_suppkey
+  );
+4095
+
+> WITH operators AS (
+  SELECT
+      dod.dataflow_id,
+      dor.id AS region_id,
+      dod.id,
+      ars.records
+  FROM
+      mz_internal.mz_dataflow_operator_dataflows dod
+      JOIN mz_internal.mz_dataflow_addresses doa
+          ON dod.id = doa.id
+      JOIN mz_internal.mz_dataflow_addresses dra
+          ON CAST(dra.address AS text) = CAST(doa.address[:list_length(doa.address) - 1] as text)
+      JOIN mz_internal.mz_dataflow_operators dor
+          ON dor.id = dra.id
+      JOIN mz_internal.mz_arrangement_sizes ars
+          ON ars.operator_id = dod.id
+  WHERE
+      dod.name = 'Arranged TopK input'
+      OR dod.name = 'Arranged MinsMaxesHierarchical input'
+      OR dod.name = 'Arrange ReduceMinsMaxes'
+  ),
+  levels AS (
+      SELECT o.dataflow_id, o.region_id, COUNT(*) AS levels
+      FROM operators o
+      GROUP BY o.dataflow_id, o.region_id
+  ),
+  pivot AS (
+      SELECT
+          o1.dataflow_id,
+          o1.region_id,
+          o1.id,
+          o1.records
+      FROM operators o1
+      WHERE
+          o1.id = (
+              SELECT MIN(o2.id)
+              FROM operators o2
+              WHERE
+                  o2.dataflow_id = o1.dataflow_id
+                  AND o2.region_id = o1.region_id
+              OPTIONS (EXPECTED GROUP SIZE = 8)
+          )
+  ),
+  candidates AS (
+      SELECT
+          o.dataflow_id,
+          o.region_id,
+          o.id,
+          o.records
+      FROM
+          operators o
+          JOIN pivot p
+              ON o.dataflow_id = p.dataflow_id
+                  AND o.region_id = p.region_id
+                  AND o.id <> p.id
+      WHERE o.records >= p.records * (1 - 0.15)
```

Comment:
Could you help me understand this part? Here we check "how many levels do not contribute to significant reductions in records maintained" I believe. I would think that requires us to compare the record counts of subsequent levels, but AFAIU this part only compares record counts with the pivot, which is the first level.

Reply:
Yes, the intent is exactly to find how many levels do not contribute to a significant drop in record count. The logic is based on the pivot because I wanted to avoid compounding the margin by only looking at adjacent levels. By comparing with the pivot, we are always contrasting with the original data that was given to the gadget.

Reply:
Maybe we can use an example to find out where my understanding fails. Suppose the following arrangement stack:

| level     | records |
|--------|---------|
| 1 (pivot) | 100 |
| 2           | 50   |
| 3           | 50   |
| 4           | 50   |
| ...          | 50   |

... i.e. after 2 levels no reduction in records takes place anymore. The query above wouldn't detect this, right? It would check for levels where the number of records is >= 85, which doesn't apply to any of the subsequent levels.

Since the query seems to work on TPC-H data, I either don't fully understand it, or I'm confused about how the record counts in the arrangement stacks can look like. Which is it?

Reply:
I see, let me try to explain how I conceptualize this. For the sake of argument, let's just say you are computing the global `MAX` and that the data are well spread out. In this case, the `EXPECTED GROUP SIZE` is the number of rows in the relation. The first level (pivot) will contain as many rows as there are in the collection (modulo compaction, etc). At each level, we use hashing to distribute the rows among groups. If you build a hierarchy that has too many levels, the first-level groups will be such that many groups might be empty or contain only one row. Each subsequent level will have a number of groups that is reduced exponentially. So at some point, we will find the level where we actually start having a few rows per group. That's where we will see the row counts drop off.

So the query is looking for this moment where the hierarchy starts having a filtering effect. Once it does, it is dangerous to suggest cutting the height of the hierarchy further. Your example above is odd, in the sense that at the first level you saw a filtering effect, so this means that the number of groups at that level was effective at achieving some reduction already. We will have way less groups in the next level, so there should be even further reduction happening or there is some substantial skew in the data. But if the latter is the case, then we should not tune the `EXPECTED GROUP SIZE` down anyway to avoid hurting latency upon updates directed at these unusually large groups.

Reply:
This makes a lot of sense, thank you!